src/drivers/rc_input/RCInput.cpp

```diff
@@ -0,0 +1,814 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "RCInput.hpp"
+
+#include "crsf_telemetry.h"
+
+using namespace time_literals;
+
+#if defined(SPEKTRUM_POWER)
+static bool bind_spektrum(int arg);
+#endif /* SPEKTRUM_POWER */
+
+work_s RCInput::_work = {};
+constexpr char const *RCInput::RC_SCAN_STRING[];
+
+RCInput::RCInput(bool run_as_task) :
+	_cycle_perf(perf_alloc(PC_ELAPSED, "rc_input cycle time")),
+	_publish_interval_perf(perf_alloc(PC_INTERVAL, "rc_input publish interval"))
+{
+	// rc input, published to ORB
+	_rc_in.input_source = input_rc_s::RC_INPUT_SOURCE_PX4FMU_PPM;
+
+	// initialize it as RC lost
+	_rc_in.rc_lost = true;
+
+	// initialize raw_rc values and count
+	for (unsigned i = 0; i < input_rc_s::RC_INPUT_MAX_CHANNELS; i++) {
+		_raw_rc_values[i] = UINT16_MAX;
+	}
+}
+
+RCInput::~RCInput()
+{
+	orb_unadvertise(_to_input_rc);
+
+	orb_unsubscribe(_adc_sub);
+	orb_unsubscribe(_vehicle_cmd_sub);
+
+#ifdef RC_SERIAL_PORT
+	dsm_deinit();
+#endif
+
+	if (_crsf_telemetry) {
+		delete (_crsf_telemetry);
+	}
+
+	perf_free(_cycle_perf);
+	perf_free(_publish_interval_perf);
+}
+
+int
+RCInput::init()
+{
+	_adc_sub = orb_subscribe(ORB_ID(adc_report));
+
+#ifdef RC_SERIAL_PORT
+
+#  ifdef RF_RADIO_POWER_CONTROL
+	// power radio on
+	RF_RADIO_POWER_CONTROL(true);
+#  endif
+	_vehicle_cmd_sub = orb_subscribe(ORB_ID(vehicle_command));
+	// dsm_init sets some file static variables and returns a file descriptor
+	_rcs_fd = dsm_init(RC_SERIAL_PORT);
```

Comment:
`_rcs_fd` is never closed.
The rest should be good, I'm ok implementing the polling in the next iteration.

Reply:
> _rcs_fd is never closed.

Never mind, `dsm_deinit` does it.